File: ios/EMWaverTests/DeviceBufferSessionTests.swift
import XCTest
@testable import EMWaver

final class DeviceBufferSessionTests: XCTestCase {
    func testSeparateSessionsKeepRxBuffersAndCountersIsolated() {
        let usb = DeviceBufferSession()
        let ble = DeviceBufferSession()

        let usbPacket = packet(0x11)
        let blePacket = packet(0x22)

        usb.storeBulkPkt(usbPacket, tsMs: 100)
        ble.storeBulkPkt(blePacket, tsMs: 200)

        XCTAssertEqual(usb.getRxPacketCount(), 1)
        XCTAssertEqual(ble.getRxPacketCount(), 1)
        XCTAssertEqual(usb.getBuffer(), usbPacket)
        XCTAssertEqual(ble.getBuffer(), blePacket)

        let usbNext = usb.nextRxPacket()
        XCTAssertEqual(usbNext?.packet64, usbPacket)
        XCTAssertNil(usb.nextRxPacket())

        let bleNext = ble.nextRxPacket()
        XCTAssertEqual(bleNext?.packet64, blePacket)
    }

    func testSeparateSessionsKeepSamplerStreamingStateIsolated() {
        let usb = DeviceBufferSession()
        let ble = DeviceBufferSession()

        _ = usb.outgoingSamplerPolicy(for: Data([0x60, 0x00]))

        let emptyLane = Data(repeating: 0x00, count: 18)
        let emptySuperframe = emptyLane + emptyLane
        guard let sysex = UsbMidiSysex.encodeSuperframe(emptySuperframe) else {
            return XCTFail("encodeSuperframe returned nil")
        }

        usb.feedMidiBytes(sysex, tsMs: 300)
        ble.feedMidiBytes(sysex, tsMs: 300)

        XCTAssertEqual(usb.getRxPacketCount(), 1)
        XCTAssertEqual(ble.getRxPacketCount(), 0)
    }

    func testSeparateSessionsKeepTxBuffersIsolated() {
        let usb = DeviceBufferSession()
        let ble = DeviceBufferSession()

        let usbPacket = packet(0x55)
        let blePacket = packet(0x66)

        usb.appendTxBytes(usbPacket, tsMs: 500)
        ble.appendTxBytes(blePacket, tsMs: 600)

        let usbTx = usb.readTxSince(packetIndex: 0, maxPackets: 8)
        let bleTx = ble.readTxSince(packetIndex: 0, maxPackets: 8)

        XCTAssertEqual(usb.getTxPacketCount(), 1)
        XCTAssertEqual(ble.getTxPacketCount(), 1)
        XCTAssertEqual(usbTx.data, Array(usbPacket))
        XCTAssertEqual(bleTx.data, Array(blePacket))
    }

    private func packet(_ value: UInt8) -> Data {
        Data(repeating: value, count: 18)
    }
}

File: windows/EMWaver.Tests/DeviceBufferSessionTests.cs
using System;
using System.Linq;
using EMWaver.Interop;
using EMWaver.Services;
using Xunit;

namespace EMWaver.Tests;

public sealed class DeviceBufferSessionTests
{
    [Fact]
    public void SeparateSessionsKeepRxBuffersAndCountersIsolated()
    {
        var usb = new DeviceBufferSession("usb:test");
        var ble = new DeviceBufferSession("ble:test");

        var usbPacket = Packet(0x11);
        var blePacket = Packet(0x22);

        usb.StoreBulkPkt(usbPacket, 100);
        ble.StoreBulkPkt(blePacket, 200);

        Assert.Equal((ulong)1, usb.GetRxPacketCount());
        Assert.Equal((ulong)1, ble.GetRxPacketCount());
        Assert.Equal(usbPacket, usb.GetRxSnapshot());
        Assert.Equal(blePacket, ble.GetRxSnapshot());

        var usbNext = usb.NextRxPacket();
        Assert.NotNull(usbNext);
        Assert.Equal(usbPacket, usbNext.Value.packet);
        Assert.Null(usb.NextRxPacket());

        var bleNext = ble.NextRxPacket();
        Assert.NotNull(bleNext);
        Assert.Equal(blePacket, bleNext.Value.packet);
    }

    [Fact]
    public void SeparateSessionsKeepSysexParserStateIsolated()
    {
        var usb = new DeviceBufferSession("usb:test");
        var ble = new DeviceBufferSession("ble:test");

        var usbCommand = Packet(0x33);
        var bleCommand = Packet(0x44);
        var emptyStream = new byte[NativeBufferRust.PacketSizeBytes];

        var usbSysex = UsbMidiSysex.EncodeSuperframe(Superframe(usbCommand, emptyStream));
        var bleSysex = UsbMidiSysex.EncodeSuperframe(Superframe(bleCommand, emptyStream));

        Assert.NotNull(usbSysex);
        Assert.NotNull(bleSysex);

        usb.FeedSysexBytes(usbSysex!, 300);
        ble.FeedSysexBytes(bleSysex!, 400);

        Assert.Equal((ulong)1, usb.GetRxPacketCount());
        Assert.Equal((ulong)1, ble.GetRxPacketCount());
        Assert.Equal(usbCommand, usb.NextRxPacket()!.Value.packet);
        Assert.Equal(bleCommand, ble.NextRxPacket()!.Value.packet);
    }

    [Fact]
    public void SeparateSessionsKeepTxBuffersIsolated()
    {
        var usb = new DeviceBufferSession("usb:test");
        var ble = new DeviceBufferSession("ble:test");

        var usbPacket = Packet(0x55);
        var blePacket = Packet(0x66);

        usb.AppendTxBytes(usbPacket, 500);
        ble.AppendTxBytes(blePacket, 600);

        Assert.Equal((ulong)1, usb.GetTxPacketCount());
        Assert.Equal((ulong)1, ble.GetTxPacketCount());
        Assert.Equal(usbPacket, usb.GetTxSnapshot());
        Assert.Equal(blePacket, ble.GetTxSnapshot());
    }

    [Fact]
    public void SysexParserAcceptsBlePaddedNotifications()
    {
        var session = new DeviceBufferSession("ble:test");
        var command = Packet(0x77);
        var emptyStream = new byte[NativeBufferRust.PacketSizeBytes];
        var sysex = UsbMidiSysex.EncodeSuperframe(Superframe(command, emptyStream));

        Assert.NotNull(sysex);
        var padded = new byte[64];
        Array.Copy(sysex!, padded, sysex!.Length);

        session.FeedSysexBytes(padded, 700);

        Assert.Equal((ulong)1, session.GetRxPacketCount());
        Assert.Equal(command, session.NextRxPacket()!.Value.packet);
    }

    private static byte[] Packet(byte value)
    {
        return Enumerable.Repeat(value, NativeBufferRust.PacketSizeBytes).ToArray();
    }

    private static byte[] Superframe(byte[] commandLane, byte[] streamLane)
    {
        var superframe = new byte[NativeBufferRust.PacketSizeBytes * 2];
        Array.Copy(commandLane, 0, superframe, 0, NativeBufferRust.PacketSizeBytes);
        Array.Copy(streamLane, 0, superframe, NativeBufferRust.PacketSizeBytes, NativeBufferRust.PacketSizeBytes);
        return superframe;
    }
}
